effect/shared/src/main/scala/scalaz/effect/KleisliIO.scala

```diff
@@ -0,0 +1,441 @@
+
+  /**
+   * Returns a new effectful function that will either compute the value of this
+   * effectful function (if passed `Left(a)`), or will compute the value of the
+   * specified effectful function (if passed `Right(c)`).
+   */
+  final def |||[C](that: KleisliIO[E, C, B]): KleisliIO[E, Either[A, C], B] =
+    KleisliIO.join(self, that)
+
+  /**
+   * Maps the output of this effectful function to the specified constant.
+   */
+  final def const[C](c: => C): KleisliIO[E, A, C] =
+    self >>> KleisliIO.lift[E, B, C](_ => c)
+
+  /**
+   * Maps the output of this effectful function to `Unit`.
+   */
+  final def toUnit: KleisliIO[E, A, Unit] = const(())
+
+  /**
+   * Returns a new effectful function that merely applies this one for its
+   * effect, returning the input unmodified.
+   */
+  final def asEffect: KleisliIO[E, A, A] = self.first >>> KleisliIO._2
+}
+
+object KleisliIO {
+  private class KleisliIOError[E](error: E) extends Throwable {
+    final def unsafeCoerce[E2] = error.asInstanceOf[E2]
+  }
+
+  private[effect] final class Pure[E, A, B](apply0: A => IO[E, B]) extends KleisliIO[E, A, B] {
+    override final def apply(a: A): IO[E, B] = apply0(a)
+  }
+  private[effect] final class Impure[E, A, B](val apply0: A => B) extends KleisliIO[E, A, B] {
+    override final def apply(a: A): IO[E, B] =
+      IO.suspend {
+        try IO.now[E, B](apply0(a))
+        catch {
+          case e: KleisliIOError[_] => IO.fail[E, B](e.unsafeCoerce[E])
+        }
+      }
+  }
+
+  /**
+   * Lifts a value into the monad formed by `KleisliIO`.
+   */
+  final def point[E, A, B](b: => B): KleisliIO[E, A, B] = lift((_: A) => b)
+
+  /**
+   * Returns a `KleisliIO` representing a failure with the specified `E`.
+   */
+  final def fail[E, A, B](e: E): KleisliIO[E, A, B] =
+    new Impure((a: A) => throw new KleisliIOError[E](e))
+
+  /**
+   * Returns the identity effectful function, which performs no effects and
+   * merely returns its input unmodified.
+   */
+  final def identity[E, A]: KleisliIO[E, A, A] = lift((a: A) => a)
+
+  /**
+   * Lifts a pure `A => IO[E, B]` into `KleisliIO`.
+   */
+  final def pure[E, A, B](f: A => IO[E, B]): KleisliIO[E, A, B] = new Pure(f)
+
+  /**
+   * Lifts a pure `A => B` into `KleisliIO`.
+   */
+  final def lift[E, A, B](f: A => B): KleisliIO[E, A, B] = new Impure(f)
+
+  /**
+   * Returns an effectful function that merely swaps the elements in a `Tuple2`.
+   */
+  final def swap[E, A, B]: KleisliIO[E, (A, B), (B, A)] =
+    KleisliIO.lift[E, (A, B), (B, A)](_.swap)
+
+  /**
+   * Lifts an impure function into `KleisliIO`, converting throwables into the
+   * specified error type `E`.
+   */
+  final def impure[E, A, B](catcher: PartialFunction[Throwable, E])(f: A => B): KleisliIO[E, A, B] =
+    new Impure(
+      (a: A) =>
+        try f(a)
+        catch {
+          case t: Throwable if (catcher.isDefinedAt(t)) =>
+            throw new KleisliIOError(catcher(t))
```

Comment:
this can be

```scala
try f(a) catch catcher.andThen(new KleisliIOError(_))
```

to avoid executing `catcher` twice.

Reply:
This doesn't type check, and I'm pretty sure even if it did, `andThen` would do the same thing behind the scenes that is being done here.